specification/datafactory/resource-manager/Microsoft.DataFactory/stable/2018-06-01/entityTypes/Dataset.json

```diff
@@ -1433,6 +1433,74 @@
         }
       }
     },
+    "AmazonRedshiftTableDataset": {
+      "x-ms-discriminator-value": "AmazonRedshiftTable",
```

Comment:
>AmazonRedshiftTable [](start = 35, length = 19)

So you're adding a new discriminator value to an existing stable API. The SDK generated from the swagger before the change will be able to deal with the discriminator value of "AmazonRedshiftTable" because it's modeled as a string, but it will not be able to deserialize the payload specific to AmazonRedshiftTable. This unfortunately makes it an SDK breaking change. Can you add this in a new API version?

Reply:
Sure. it will be generated in a new API version when making changes to SDK.

Reply:
as similar as this SDK change where you can see that we change the API version https://github.com/Azure/azure-sdk-for-net/pull/7044

Reply:
To generate a new API version when making changes to SDK, you need the swagger to exist for the new API version, no?

---
In reply to: [310406114](https://github.com/Azure/azure-rest-api-specs/pull/6820#discussion_r310406114) [](ancestors = 310406114)

Reply:
right. the order I understand is also the same as you : merge swagger change first, then generate SDK based on the latest swagger.

Reply:
What confused me a little bit from your concern is: AmazonRedshiftTableDataset is a new dataset why it is blocker or breaking change here?

Reply:
if there is any chance you check the history of this file, you will see that we keep adding different datasets as similar as this one..

Reply:
If there is already an SDK that was generated from swagger API version 2018-06-01 and a customer is using it, then this particular SDK will not be able to handle a new response for the new enum value and new properties. 

---
In reply to: [310745770](https://github.com/Azure/azure-rest-api-specs/pull/6820#discussion_r310745770) [](ancestors = 310745770)

Reply:
Repeated a bad pattern is not a good thing. The real question here is how do you ensure that existing SDK customers switch over to the new SDK that can handle the new feature. If you have an existing way of mitigating that, then we might be able to approve. If not, you would need to introduce a new API version.

---
In reply to: [310746499](https://github.com/Azure/azure-rest-api-specs/pull/6820#discussion_r310746499) [](ancestors = 310746499)

Reply:
thanks for your comments. @majastrz. Our existing pattern is by design like this. We keep adding new connectors which have new linked service, new datasets, new source which are all added into the same Swagger version. at the meantime, we release SDKs with newer versions from time to time.  here is our latest SDK where you can see the detailed versions: https://www.nuget.org/packages/Microsoft.Azure.Management.DataFactory/4.1.1
also cc @hvermis  here.

Reply:
How do your customers avoid breakage when a new connector is added?

Reply:
From SDK perspective? every new connector is a separate one and not built upon the existing ones. If they need to leverage the new connector, they use a new version of SDK.

Reply:
So what prevents a user using an older SDK versions from executing a GET collection and seeing resources created via a new SDK using new connectors?

Reply:
I assume using old SDK version cannot see the new connectors which are only included in the new SDK, this is also the reason why they need to upgrade their SDK if they need to leverage the new connectors.

Reply:
the new SDK version is always built based on the latest swagger. which means the old SDK were built based on swagger at specific time and then kept intact there.

Reply:
Let's not assume. Can you verify the behavior? What I want to know if it will throw. If the SDK just accepts the payload even if the user can't really interact with it, I'd be ok with that.

@michaeljqzq, can you chime in here as well? What is the behavior with the generated SDK if a new discriminator value is being returned in the backend, but the SDK doesn't know about it yet? Will it throw an exception?

Reply:
I double checked our changes on SDK and can confirm that we manually changed the version and appending the new changes on existing one every time to the new version. @hvermis could you please also help confirm as well?  thanks.

Reply:
We do not update the API version. This is a known behavior and we have been adding new types like this. Our customers are aware of this.

Reply:
@wenbof and @hvermis, you are not actually answering my question, which is delaying this review. I'm not asking whether you have done this before (you clearly have). Azure is a distributed system. There's no mechanism to ensure that all customers simultaneously upgrade to latest and greatest SDK that you have generated, which is why we have breaking changes guidance.

So I will ask the question again in a different way. Let's say you have SDK v2 that supports connectors a and b and SDK v3 that supports connectors a, b, and c. User A has been happily using SDK v2 to manage resources using connectors a and b. Suddenly user B creates resources using connector c in the same subscription or RG via SDK v3. 

What will User A now see when he/she queries for all resources?
1. Will the old SDK v2 throw an exception?
2. Will the responses deserialize successfully but incompletely?

Please pick either answer 1 or 2.